### PR TITLE
Test old client versions work with the latest service version.

### PR DESCRIPTION
**Pull request by LukeButters**
# Background

Adds support for testing old clients communicating with new services.

The setup is best described in this image:

Testing old halibut client@2x (2)

When Testing an old client with a new services, those tests will also happens to test new client to old service.

[SC-46260]

Since the client is in an external process, it is misleading to use `CreateClient` as the method name since configuration on that client wont work as expected e.g. a `CancellationToken` will be the CT on the top arrow of the diagram, while the test is mostly likely wanting to set that on the CT for the bottom arrow communications.

The result of the build method implements `IClientAndService`, but some methods throw since they don't make sense.

# Other changes

Also adds logging to Halibut within the external binaries to make debugging easier, for example debugging the development of this feature.

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

**Comment by nathanwoctopusdeploy** [comment]
[sc-53161]

File: source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+            
+            // Create a client which will send RPC calls to the service in the Test CLR.
+            Console.WriteLine("This should be bob:");
+            Console.WriteLine(clientCert.ToString());
+            using (var client = new HalibutRuntimeBuilder()
+                       .WithServerCertificate(clientCert)
+                       .WithLogFactory(new TestContextLogFactory("ProxyClient"))
+                       .Build())
+            {
+                //
```
Comment (sburmanoctopus) [inline source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs:57]: //?

File: source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    /// 
+    /// This is used when testing the compatibility of a old client with latest service, since this
+    /// has the old client which is making requests to the new service in the test CLR.
+    /// The Proxy 
+    /// </summary>
+    public class ProxyServiceForwardingRequestToClient
+    {
+        public static int Run(string[] args)
+        {
+            var tentacleCertPath = Environment.GetEnvironmentVariable("tentaclecertpath");
```
Comment (sburmanoctopus) [inline source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs:21]: Nit: this method is a bit hard to follow. Any chance of breaking it up a bit?

File: source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                            proxyService.Trust(clientCert.Thumbprint);
+                            break;
+                        default:
+                            throw new ArgumentOutOfRangeException();
+                    }
+
+                    Console.WriteLine("RunningAndReady");
+                    Console.WriteLine("Will Now sleep");
+                    Console.Out.Flush();
+                    Thread.Sleep(1000000);
```
Comment (sburmanoctopus) [inline source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs:117]: What's the sleep for?

File: source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs (hunk 1 above)
Comment (LukeButters) [reply source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs]: Removed

File: source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs (hunk 3 above)
Comment (LukeButters) [reply source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs:117]: Add a note on what is going on here.

File: source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs (hunk 2 above)
Comment (LukeButters) [reply source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs:21]: Yeah its a bit long, it can be broken up as the changes need to occur to this method.

**Comment by nathanwoctopusdeploy** [comment]
💭 For another PR

Could we work around the test limitations by adding "smarts" to the Proxy Halibut Service e.g. add a new Service `CancelCancellationToken[ForRequest]`, then wrapping the Test Halibut Client to capture cancellation tokens and call the additional services `CancelCancellationToken[ForRequest]` to "flow cancellation". The Halibut Proxy Service would have to track tokens and requests etc so would have a bit of logic!!

File: source/Halibut.TestUtils.CompatBinary.v5_0_429/Program.cs [hunk 4]
```diff
@@ -7,6 +8,8 @@ public class Program
     {
         public static int Main(string[] args)
         {
+            var s = HalibutLimits.PollingRequestQueueTimeout;
```
Comment (nathanwoctopusdeploy) [inline source/Halibut.TestUtils.CompatBinary.v5_0_429/Program.cs:11]: ??

File: source/Halibut.Tests/BackwardsCompatibility/TestOldClientWithNewService.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+using NUnit.Framework;
+
+namespace Halibut.Tests.BackwardsCompatibility
+{
+    public class TestOldClientWithNewService
+    {
+        [Test]
+        public async Task Listening()
+        {
+            new SerilogLoggerBuilder().Build().Information("Hello");
```
Comment (nathanwoctopusdeploy) [inline source/Halibut.Tests/BackwardsCompatibility/TestOldClientWithNewService.cs:17]: ??

File: source/Halibut.Tests/BackwardsCompatibility/TestOldClientWithNewService.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+            var echoService = new CallRecordingEchoServiceDecorator(new EchoService());
+            using (var clientAndService = await PreviousClientVersionAndServiceBuilder.WithListeningService().WithClientVersion(PreviousServiceVersions.v5_0_429)
```
Comment (nathanwoctopusdeploy) [inline source/Halibut.Tests/BackwardsCompatibility/TestOldClientWithNewService.cs:19]: `WithClientVersion(PreviousServiceVersions.v5_0_429)` is a little confusing - probably want to rename `PreviousServiceVersions` to something more generic

File: source/Halibut.Tests/BackwardsCompatibility/TestOldClientWithNewService.cs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+            echoService.SayHelloCallCount.Should().Be(1);
+        }
+        
+        [Test]
+        public async Task Polling()
+        {
+            new SerilogLoggerBuilder().Build().Information("Hello");
+            var echoService = new CallRecordingEchoServiceDecorator(new EchoService());
+            using (var clientAndService = await PreviousClientVersionAndServiceBuilder.WithPollingService().WithClientVersion(PreviousServiceVersions.v5_0_429)
```
Comment (nathanwoctopusdeploy) [inline source/Halibut.Tests/BackwardsCompatibility/TestOldClientWithNewService.cs:36]: Can these 2 tests become 1 with an TestCaseSource to control Polling/Listening so were not repeating test code for all ServiceTypes

File: source/Halibut.Tests/BackwardsCompatibility/TestOldClientWithNewService.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+                echo.SayHello("hello");
+            }
+
+            echoService.SayHelloCallCount.Should().Be(1);
+        }
+        
+        [Test]
+        public async Task Polling()
+        {
+            new SerilogLoggerBuilder().Build().Information("Hello");
```
Comment (nathanwoctopusdeploy) [inline source/Halibut.Tests/BackwardsCompatibility/TestOldClientWithNewService.cs:34]: ??

File: source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+                
+                Uri realServiceUri = null;
+                switch (serviceConnectionType)
+                {
+                    case ServiceConnectionType.Polling:
+                        var clientPollingListeningPort = client.Listen();
+                        Console.WriteLine("Polling listener is listening on port: " + clientPollingListeningPort);
+                        
+                        // This needs to be done within the CLR.
```
Comment (nathanwoctopusdeploy) [inline source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs:69]: ??

File: source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+                        break;
+                    default:
+                        throw new ArgumentOutOfRangeException();
+                }
+
+                var realServiceEndpoint = new ServiceEndPoint(realServiceUri, serviceCert.Thumbprint);
+
+                var forwardingEchoService = client.CreateClient<IEchoService>(realServiceEndpoint);
+                
+                // We can only test in listening mode since in polling not everything is setup yet.
```
Comment (nathanwoctopusdeploy) [inline source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs:84]: ??

File: source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs (hunk 2 above)
Comment (nathanwoctopusdeploy) [reply source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs:21]: +1- could do with splitting up for clarity

File: source/Halibut.Tests/Support/BackwardsCompatibility/PreviousClientVersionAndServiceBuilder.cs [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+        public async Task<ClientAndService> Build()
+        {
+            
+            // We need to build both a Client and a service.
+            if (version == null) throw new Exception("The version of the service must be set.");
+            var client = new HalibutRuntime(clientCertAndThumbprint.Certificate2);
+            client.Trust(serviceCertAndThumbprint.Thumbprint);
+
+            var realService2 = ClientServiceBuilder.ForMode(serviceConnectionType)
```
Comment (nathanwoctopusdeploy) [inline source/Halibut.Tests/Support/BackwardsCompatibility/PreviousClientVersionAndServiceBuilder.cs:77]: This is not used

**Review by nathanwoctopusdeploy — APPROVED** [review]
Looks good. Solves a tricky problem in a clever way.

My feedback would be around how understandable and maintainable it's going to be. You need to understand all the moving pieces, and there are quite a few. Engineers could struggle to debug problems due to the complexity of jumping across processes and the non-trivial setup required to achieve this.

I don't think the approach needs to change, but I think the code needs to be revisited so it's really clear what is happening. Refactor to break up the pieces (just split the big long methods apart) / really good comments explaining what is happening to the reader.

Feel free to merge this and revisit it as tech debt, but a "must have".

File: source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs (hunk 9 above)
Comment (LukeButters) [reply source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs:69]: I think they are notes for me when I was converting it. I will remove.

File: source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs (hunk 10 above)
Comment (LukeButters) [reply source/Halibut.TestUtils.CompatBinary.Base/ProxyServiceForwardingRequestToClient.cs:84]: I think removing makes sense since testing means extra calls, which could interfere with tests.

File: source/Halibut.Tests/BackwardsCompatibility/TestOldClientWithNewService.cs (hunk 5 above)
Comment (LukeButters) [reply source/Halibut.Tests/BackwardsCompatibility/TestOldClientWithNewService.cs:17]: No it is just courteous. :laughing:

File: source/Halibut.Tests/BackwardsCompatibility/TestOldClientWithNewService.cs (hunk 6 above)
Comment (LukeButters) [reply source/Halibut.Tests/BackwardsCompatibility/TestOldClientWithNewService.cs:19]: good idea, renamed to `PreviousVersions`

File: source/Halibut.Tests/BackwardsCompatibility/TestOldClientWithNewService.cs (hunk 7 above)
Comment (LukeButters) [reply source/Halibut.Tests/BackwardsCompatibility/TestOldClientWithNewService.cs]: Will do once https://github.com/OctopusDeploy/Halibut/pull/270 is in. Doing it now will cause a poison merge since websockets are not supported yet.
